multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java

```diff
@@ -0,0 +1,177 @@
+            } finally {
+                bc.abort();
+            }
+            return null;
+        }
+
+    }
+
+    @Extension public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        public DescriptorImpl() {
+            super(Execution.class);
+        }
+
+        @Override public String getFunctionName() {
+            return "properties";
+        }
+
+        @Override public String getDisplayName() {
+            return "Set job properties";
+        }
+
+        @Override public Step newInstance(StaplerRequest req, JSONObject formData) throws FormException {
+            // A modified version of RequestImpl.TypePair.convertJSON.
+            // Works around the fact that Stapler does not call back into Descriptor.newInstance for nested objects.
+            List<JobProperty> properties = new ArrayList<JobProperty>();
+            ClassLoader cl = req.getStapler().getWebApp().getClassLoader();
+            @SuppressWarnings("unchecked") Set<Map.Entry<String,Object>> entrySet = formData.getJSONObject("propertiesMap").entrySet();
+            for (Map.Entry<String,Object> e : entrySet) {
+                if (e.getValue() instanceof JSONObject) {
+                    String className = e.getKey().replace('-', '.'); // decode JSON-safe class name escaping
+                    Class<? extends JobProperty> itemType;
+                    try {
+                        itemType = cl.loadClass(className).asSubclass(JobProperty.class);
+                    } catch (ClassNotFoundException x) {
+                        throw new FormException(x, "propertiesMap");
+                    }
+                    JobPropertyDescriptor d = (JobPropertyDescriptor) Jenkins.getActiveInstance().getDescriptorOrDie(itemType);
+                    JSONObject more = (JSONObject) e.getValue();
+                    JobProperty property = d.newInstance(req, more);
+                    if (property != null) {
+                        properties.add(property);
+                    }
+                }
+            }
+            return new JobPropertyStep(properties);
+        }
+
+        @Restricted(DoNotUse.class) // f:repeatableHeteroProperty
+        public Collection<? extends Descriptor<?>> getPropertyDescriptors() {
+            List<Descriptor<?>> result = new ArrayList<Descriptor<?>>();
+            for (JobPropertyDescriptor p : ExtensionList.lookup(JobPropertyDescriptor.class)) {
+                if (p.isApplicable(WorkflowJob.class)) {
+                    result.add(p);
+                }
+            }
+            return result;
+        }
+
+    }
+
+    @Extension public static class HideSuperfluousBranchProperties extends DescriptorVisibilityFilter {
+
+        @Override public boolean filter(Object context, Descriptor descriptor) {
+            if (context instanceof WorkflowMultiBranchProject && (descriptor.clazz == RateLimitBranchProperty.class || descriptor.clazz == BuildRetentionBranchProperty.class)) {
```

Comment:
@amuniz please comment on the upstream PR except for things specifically added in this PR.